pkg/apis/pingcap/v1alpha1/tidbcluster.go

```diff
@@ -13,6 +13,177 @@
 
 package v1alpha1
 
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// defaultHelperImage is default image of helper
+	defaultHelperImage = "busybox:1.26.2"
+)
+
+// ComponentAccessor is the interface to access component details, which respects the cluster-level properties
+// and component-level overrides
+type ComponentAccessor interface {
+	Image() string
+	ImagePullPolicy() corev1.PullPolicy
+	HostNetwork() bool
+	Affinity() *corev1.Affinity
+	PriorityClassName() string
+	NodeSelector() map[string]string
+	Annotations() map[string]string
+	Tolerations() []corev1.Toleration
+	PodSecurityContext() *corev1.PodSecurityContext
+	SchedulerName() string
+}
+
+type componentAccessorImpl struct {
+	// Cluster is the TidbCluster Spec
+	ClusterSpec *TidbClusterSpec
+
+	// Cluster is the Component Spec
+	ComponentSpec *ComponentSpec
+}
+
+func (a *componentAccessorImpl) Image() string {
+	image := a.ComponentSpec.Image
+	baseImage := a.ComponentSpec.BaseImage
+	// base image takes higher priority
+	if baseImage != "" {
+		version := a.ComponentSpec.Version
+		if version == "" {
+			version = a.ClusterSpec.Version
+		}
+		image = fmt.Sprintf("%s:%s", baseImage, version)
+	}
+	return image
+}
+
+func (a *componentAccessorImpl) PodSecurityContext() *corev1.PodSecurityContext {
+	return a.ComponentSpec.PodSecurityContext
+}
+
+func (a *componentAccessorImpl) ImagePullPolicy() corev1.PullPolicy {
+	pp := a.ComponentSpec.ImagePullPolicy
+	if pp == nil {
+		pp = &a.ClusterSpec.ImagePullPolicy
+	}
```

Comment:
`ClusterSpec.ImagePullPolicy` must not be empty?

Reply:
Yes, there will be a defaulting through mutating-webhook and validation to ensure clusterSpec.ImagePullPolicy and component.ImagePullPolicy is not both empty.

Reply:
And it's not an issue if clusgerSpec.ImagePullPolicy is empty, we just write empty pull policy in PodSpec and use the default of kubernetes